### PR TITLE
Insert empty token for empty macro arguments.

### LINKED ISSUE #27: Failure from math.h on Linux: macro invocations with empty parameter
@wvdschel reported that he is having a melange failure on Linux when processing ``math.h``.

The problematic code can be reduced to:

```c
#define __MATHCALL(function,suffix, args)       \
  __MATHDECL (_Mdouble_,function,suffix, args)
#define __MATHDECL(type, function,suffix, args) \
  __MATHDECL_1(type, function,suffix, args); \
  __MATHDECL_1(type, __CONCAT(__,function),suffix, args)
#define __MATHCALLX(function,suffix, args, attrib)      \
  __MATHDECLX (_Mdouble_,function,suffix, args, attrib)
#define __MATHDECLX(type, function,suffix, args, attrib) \
  __MATHDECL_1(type, function,suffix, args) __attribute__ (attrib); \
  __MATHDECL_1(type, __CONCAT(__,function),suffix, args) __attribute__ (attrib)
#define __MATHDECL_1(type, function,suffix, args) \
  extern type __MATH_PRECNAME(function,suffix) args __THROW
#define _Mdouble_               double
#define __MATH_PRECNAME(name,r) __CONCAT(name,r)
#define __CONCAT(x,y)   x ## y
#define __THROW

__MATHCALL (acos,, (_Mdouble_ __x));
```

Specifically, the problem is with the empty parameter in ``__MATHCALL (acos,, (_Mdouble_ __x));``.

### PR DESCRIPTION
**Pull request by wvdschel**
This prevents the concatenation operator (##) from erroneously
consuming the next token when trying to concatenate an empty
macro argument with some prefix.

Needed for parsing math.h/mathcalls.h on Linux, fixes #27.

### CODE REVIEW

File: tests/test.empty-macro-arg/expected.dylan [hunk 1]
```diff
@@ -0,0 +1,21 @@
+module: test-empty-macro-arg
+synopsis: minimal test case for melange bug when passing empty macro args
+author: Wim Vander Schelden
+copyright: X11
```
Comment (waywardmonkeys) [inline tests/test.empty-macro-arg/expected.dylan]: can you drop this line and leave it as implied the same as the rest of the code? or explicitly state that it is licensed according to the License.txt as other files do? (same wording)

File: tests/test.empty-macro-arg/test.intr [hunk 2]
```diff
@@ -0,0 +1,12 @@
+module: test-empty-macro-arg
+synopsis: minimal test case for melange bug when passing empty macro args
+author: Wim Vander Schelden
+copyright: X11
```
Comment (waywardmonkeys) [inline tests/test.empty-macro-arg/test.intr]: As above...

File: melange-core/c-lexer-cpp.dylan [hunk 3]
```diff
@@ -188,7 +193,7 @@ define /* exported */ function check-cpp-expansion
      current-depth = 0)
  => (result :: <boolean>);
   let headless-string
-    = if (string.first == '#') copy-sequence(string, start: 1) else string end;
+    = if (string.size > 0 & string.first == '#') copy-sequence(string, start: 1) else string end;
```
Comment (waywardmonkeys) [inline melange-core/c-lexer-cpp.dylan]: You can use ``~string.empty?``

File: tests/test.empty-macro-arg/empty-macro-arg.h [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#define MANGLE(module, name) module ## name
+#define DEFUN(name, module, args) extern void MANGLE(module, name) args
+
+DEFUN(foo,, (int num));
+DEFUN(bar, stdlib, (int num));
+
+#define __MATHCALL(function,suffix, args) extern double function ## suffix args 
+__MATHCALL (acos,, (int __x));
```
Comment (waywardmonkeys) [inline tests/test.empty-macro-arg/empty-macro-arg.h]: Can you add a test that has multiple empty tokens?

File: melange-core/c-lexer-cpp.dylan [hunk 5]
```diff
@@ -224,7 +229,7 @@ define /* exported */ function check-cpp-expansion
                   string);
     member?(string, forbidden-expansions, test: \=) =>
       #f;
-    string.first == '#' =>
+    string.size > 0 & string.first == '#' =>
```
Comment (waywardmonkeys) [inline melange-core/c-lexer-cpp.dylan]: ``~string.empty?`` here too.

File: melange-core/c-lexer-cpp.dylan [hunk 6]
```diff
@@ -130,7 +131,11 @@ define function get-macro-params
       <rparen-token> => paren-count := paren-count - 1;
       otherwise => #f;
     end select;
+    empty-token := #f
   finally
+    if (empty-token) // Insert an empty token.
+      list := pair(make(<identifier-token>, string: "", generator: state), list)
+    end if;
```
Comment (waywardmonkeys) [inline melange-core/c-lexer-cpp.dylan]: Maybe I'm missing something ... it looks like this is just adding a single empty token to the ``list`` at head of the list (since the list is in reverse order). Why is that the right solution? Shouldn't there be an empty token in the right spot in the list?

Comment (wvdschel) [inline melange-core/c-lexer-cpp.dylan]: The for loop above (starting at line 123) recursively adds tokens at the head of the list, until it determines the end of the list has been reached.

If no token was inserted in that loop, I add the empty token at the head of the list, where the argument would have been.

**Comment by waywardmonkeys** [comment]
I need a bit of time to sort through this and understand it. Remind me if I haven't responded within a day or so.

**Comment by wvdschel** [comment]
The main change I made over the first commit was making `get-token` skip all empty identifier tokens, unless a new keyword argument (`skip-empty-tokens`) is specifically set to true. I currently only pass this argument in the code where the concatenation operator `##` is implemented.

This works for all the test cases I have tried, though obviously it may still mishandle some cases I haven't thought off.
